### ISSUE TITLE
Mistral slow shutdown

### ISSUE THREAD

**Issue by nmaludy** [StackStorm/st2mistral#28]
Over the past few weeks i've noticed `st2ctl stop` takes a fairly long time. Doing a little digging tonight and i realized that the mistral service is the one that takes the longest to shut down.

I decided to try and stop the service manually using `systemctl stop mistral-server` and it still took a long time. In fact it took 1 minute:

```
[root@stackstorm ~]# time systemctl stop mistral

real    1m0.024s
user    0m0.002s
sys     0m0.006s
```

Doing some grepping i found the configuration parameter `rpc_response_timeout`: https://github.com/StackStorm/mistral/blob/master/mistral/config.py#L105

Changing this value to something _other than_ 60 results in a _much_ quicker shutdown:

With a value of 15:
```
[root@stackstorm ~]# time systemctl stop mistral

real    0m1.024s
user    0m0.003s
sys     0m0.007s
```

With a value of 30:
```
[root@stackstorm ~]# time systemctl stop mistral

real    0m5.164s
user    0m0.002s
sys     0m0.008s
```

With a value of 61 thanks @bigmstone :
```
[root@stackstorm ~]# time systemctl stop mistral

real    0m14.093s
user    0m0.005s
sys     0m0.009s
```

And again with a value of 60:
```
root@stackstorm ~]# time systemctl stop mistral

real    1m0.040s
user    0m0.006s
sys     0m0.006s
```

It looks like the only place this value is used is here: https://github.com/StackStorm/mistral/blob/master/mistral/engine/rpc_backend/kombu/kombu_client.py#L108

And the underlying data structure that's being used is a `six.moves.queue.Queue` here https://github.com/StackStorm/mistral/blob/master/mistral/engine/rpc_backend/kombu/kombu_listener.py#L42

The whole goal of this issue is simply to improve the shutdown time of the mistral service (without degrading performance or data consistency of course).

**Comment by armab** [comment]
Thanks for the research!

A bit related to another https://github.com/StackStorm/st2-packages/pull/464 where guys experienced slow mistral shutdown too.

If @m4dcoder agrees that lowering default `rpc_response_timeout` is a good idea, - we can change `mistral.conf` https://github.com/StackStorm/st2-packages/blob/cec7c0424e3787f6d17bf5b637d97b6eb512b72f/packages/st2mistral/conf/mistral.conf#L167-L168 bundled in `st2mistral` package.

**Comment by bigmstone** [comment]
@armab I'm all for the quick fix, but would also love to get to the bottom of this too as to why `60` seconds is the magic slow down number. IMO we should keep this open until we work out why `60` seems mystic.

**Comment by armab** [comment]
Changing the `rpc_response_timeout` in `mistral.conf` didn't help me under Ubuntu Xenial.
`service mistral stop` still takes `30s` to complete:

```
root@ubuntu16:~# time service mistral stop

real	0m30.285s
user	0m0.004s
sys	0m0.000s

```

cc @m4dcoder would be nice to investigate what's going on behind the hood, since mistral stop is slowing down entire `st2ctl`.

**Comment by skjbulcher** [comment]
Altering the `rpc_response_timeout` does not appear to affect the docker official docker container either. I'm running a base configuration [docker-compose.yml](https://github.com/StackStorm/st2-docker/blob/eb0beb5c60a191d15cb920588818e8c30305a255/docker-compose.yml) with chatops configured. I made several change to the `rpc_response_timeout` value, with no significant impact on the 'real' time, which you see below:

```
# time st2ctl restart
st2actionrunner stop/waiting
st2api stop/waiting
st2stream stop/waiting
st2auth stop/waiting
st2garbagecollector stop/waiting
st2notifier stop/waiting
st2resultstracker stop/waiting
st2rulesengine stop/waiting
st2sensorcontainer stop/waiting
st2chatops stop/waiting
mistral stop/waiting
st2actionrunner start/running
st2api start/running, process 12045
st2stream start/running, process 12089
st2auth start/running, process 12120
st2garbagecollector start/running, process 12147
st2notifier start/running, process 12154
st2resultstracker start/running, process 12161
st2rulesengine start/running, process 12168
st2sensorcontainer start/running, process 12175
st2chatops start/running, process 12182
mistral start/running
##### st2 components status #####
st2actionrunner PID: 12031
st2actionrunner PID: 12034
st2actionrunner PID: 12037
st2api PID: 12045
st2api PID: 12075
st2stream PID: 12089
st2stream PID: 12111
st2auth PID: 12120
st2auth PID: 12136
st2garbagecollector PID: 12147
st2notifier PID: 12154
st2resultstracker PID: 12161
st2rulesengine PID: 12168
st2sensorcontainer PID: 12175
st2chatops PID: 12182
mistral-server PID: 12401
mistral-api PID: 12393

real    1m20.893s
user    0m0.000s
sys     0m0.016s
```

**Comment by armab** [comment]
There is a patch submitted in upstream mistral to address this issue: https://review.openstack.org/#/c/499853

**Comment by armab** [comment]
This is fixed in mistral upstream and included in StackStorm 2.4.1 https://stackstorm.com/2017/09/18/quick-one-st2-2-4-1-released/
